File: 2018/02/02.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <array>
#include <string>
#include <iterator>
#include <algorithm>

using namespace std;

int main()
{
    std::fstream input( "input.txt", std::ios::out | std::ios::in );
    
    vector<string> input_arr;
    string tmp;

    int two = 0;
    int three = 0;

    while ( input >> tmp )
    {
        input_arr.push_back( tmp );
    }

    //first part
    for ( auto& str : input_arr )
    {
        array<int, 26> letters{ 0 };
        
        for ( auto& letter : str )
        {
            letters.at( letter - 'a' )++;
        }

        auto pos = find( begin( letters ), end( letters ), 2 );

        if ( *pos == 2 )
        {
            two++;
        }

        pos = find( begin( letters ), end( letters ), 3 );

        if ( *pos == 3 )
        {
            three++;
        }
    }

    cout << two * three << endl;

    //second part
    for ( auto str = begin( input_arr ); str != end( input_arr ); ++str )
    {
        for ( auto str2 = str + 1; str2 != end( input_arr ); ++str2 )
        {
            string common = *str;

            for ( auto fir = begin( *str ), sec = begin( *str2 ); fir != end( *str ); ++fir, ++sec )
            {
                if ( *fir != *sec )
                {
                    common.erase( common.begin() + 1 );
                }
            }

            if ( common.size() == str->size() - 1 )
            {
                cout << common << endl;
            }
        }
    }



    return 0;
}

File: 2018/02/02.kt
package pl.jaremko.advent

import java.io.File

fun main(args: Array<String>) {
    val inputFile = File("input.txt")
    var inputArr = ArrayList<String>()

    var two = 0
    var three = 0

    inputFile.forEachLine{ inputArr.add( it ) }

    for ( str in inputArr ) {
        var letters = IntArray( 26 )

        for ( letter in str ) {
            letters[ letter.toInt() - 'a'.toInt() ]++
        }

        if ( letters.contains( 2 ) ) {
            two++
        }

        if ( letters.contains( 3 ) ) {
            three++
        }
    }

    println( two * three )

    for ( i in 0..inputArr.size - 1 ) {
        for ( j in i + 1..inputArr.size - 1 ) {
            var common = inputArr[ i ]
            var first = 0
            var second = 0

            while ( first < inputArr[ i ].length ) {
                if ( inputArr[ i ][ first ] != inputArr[ j ][ second ] ) {
                    common = common.removeRange( 0, 1 );
                }

                first++
                second++
            }

            if ( common.length == inputArr[ i ].length - 1 ) {
                println( common )
            }
        }
    }
}
